File: tt.py
import os
import yt_dlp
from telegram import Update
from telegram.ext import Application, CommandHandler, MessageHandler, filters
from telegram.ext import CallbackContext

# Ganti dengan token bot Telegram Anda
BOT_TOKEN = "7846302722:AAF_UHsrR0KDKIX16l4XI5rStgDiXWyVA8c"

# Folder untuk menyimpan file sementara
DOWNLOAD_FOLDER = "./downloads/"

# Pastikan folder download ada
os.makedirs(DOWNLOAD_FOLDER, exist_ok=True)

# Fungsi untuk menangani perintah '/start'
async def start(update: Update, context: CallbackContext):
    await update.message.reply_text(
        "Selamat datang! Kirim URL video dari TikTok, Instagram, atau YouTube untuk diunduh."
    )

# Fungsi untuk menangani pengunduhan video dari TikTok, Instagram, atau YouTube
async def download_video(update: Update, context: CallbackContext):
    url = update.message.text.strip()
    if not any(domain in url for domain in ['tiktok.com', 'instagram.com', 'youtube.com']):
        await update.message.reply_text("Silakan kirim URL TikTok, Instagram, atau YouTube yang valid.")
        return

    try:
        # Set konfigurasi unduhan untuk video
        ydl_opts = {
            'format': 'bestvideo[height<=1080]+bestaudio/best',  # Resolusi maksimal 1080p
            'merge_output_format': 'mp4',  # Menggabungkan video dan audio ke format MP4
            'outtmpl': os.path.join(DOWNLOAD_FOLDER, '%(id)s.%(ext)s'),  # Penamaan file
        }

        # Unduh video dari URL
        with yt_dlp.YoutubeDL(ydl_opts) as ydl:
            info_dict = ydl.extract_info(url, download=True)
            video_path = os.path.join(DOWNLOAD_FOLDER, f"{info_dict['id']}.mp4")

            # Kirim file video ke pengguna
            await update.message.reply_video(video=open(video_path, 'rb'))

            # Hapus file setelah dikirim
            os.remove(video_path)
    except Exception as e:
        await update.message.reply_text(f"Terjadi kesalahan: {e}")

# Fungsi utama untuk menjalankan bot
def main():
    # Buat aplikasi bot
    application = Application.builder().token(BOT_TOKEN).build()

    # Daftarkan handler untuk perintah dan pesan
    application.add_handler(CommandHandler("start", start))
    application.add_handler(MessageHandler(filters.TEXT & ~filters.COMMAND, download_video))

    # Jalankan bot
    application.run_polling()

if __name__ == '__main__':
    main()

File: tt.js
const { Telegraf } = require('telegraf');
const ytdl = require('ytdl-core');
const TikTokScraper = require('tiktok-scraper');
const axios = require('axios');
const fs = require('fs');
const path = require('path');

// Token Telegram Bot
const BOT_TOKEN = "7561651517:AAGzQ_xaWcoEHZXy27G5OyHP9UFJn0e01Ms";
const bot = new Telegraf(BOT_TOKEN);

// Folder untuk menyimpan file sementara
const DOWNLOAD_FOLDER = './downloads/';
if (!fs.existsSync(DOWNLOAD_FOLDER)) {
    fs.mkdirSync(DOWNLOAD_FOLDER);
}

// Fungsi untuk memulai bot
bot.start((ctx) => {
    ctx.reply("Halo! Kirim tautan YouTube, TikTok, atau Instagram untuk mengunduh audionya saja.");
});

// Fungsi untuk mengunduh audio
bot.on('text', async (ctx) => {
    const url = ctx.message.text.trim();

    // Validasi URL
    if (!url.includes('youtube.com') && !url.includes('youtu.be') && !url.includes('tiktok.com') && !url.includes('instagram.com')) {
        ctx.reply("Kirim link dari YouTube, TikTok, atau Instagram yang valid.");
        return;
    }

    ctx.reply("Sedang memproses audio, harap tunggu...");

    try {
        let filePath;

        if (url.includes('youtube.com') || url.includes('youtu.be')) {
            // Unduh audio dari YouTube
            const info = await ytdl.getInfo(url);
            const title = info.videoDetails.title.replace(/[\/\\?%*:|"<>]/g, '-');
            filePath = path.join(DOWNLOAD_FOLDER, `${title}.mp3`);
            const audioStream = ytdl(url, { filter: 'audioonly', quality: 'highestaudio' });

            // Simpan audio ke file
            const writeStream = fs.createWriteStream(filePath);
            audioStream.pipe(writeStream);
            await new Promise((resolve) => writeStream.on('finish', resolve));

        } else if (url.includes('tiktok.com')) {
            // Unduh audio dari TikTok
            const videoMeta = await TikTokScraper.getVideoMeta(url);
            const audioUrl = videoMeta.collector[0].musicUrl;
            const title = videoMeta.collector[0].text.replace(/[\/\\?%*:|"<>]/g, '-');
            filePath = path.join(DOWNLOAD_FOLDER, `${title}.mp3`);

            // Unduh dan simpan audio
            const response = await axios({
                url: audioUrl,
                method: 'GET',
                responseType: 'stream',
            });
            const writeStream = fs.createWriteStream(filePath);
            response.data.pipe(writeStream);
            await new Promise((resolve) => writeStream.on('finish', resolve));

        } else if (url.includes('instagram.com')) {
            ctx.reply("Instagram saat ini tidak mendukung pengunduhan audio saja.");
            return;
        }

        // Kirim file audio ke pengguna
        if (fs.existsSync(filePath)) {
            await ctx.replyWithAudio({ source: filePath });
            fs.unlinkSync(filePath); // Hapus file setelah dikirim
        } else {
            throw new Error("File audio tidak ditemukan setelah diunduh.");
        }
    } catch (err) {
        console.error(err);
        ctx.reply(`Terjadi kesalahan: ${err.message}`);
    }
});

// Jalankan bot
bot.launch().then(() => {
    console.log("Bot berjalan...");
});

// Graceful shutdown
process.once('SIGINT', () => bot.stop('SIGINT'));
process.once('SIGTERM', () => bot.stop('SIGTERM'));
